File: Java/app/src/main/java/kz/sgq/fs_tiktaktoe/mvp/presenter/FieldPresenter.java
package kz.sgq.fs_tiktaktoe.mvp.presenter;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;

import kz.sgq.fs_tiktaktoe.mvp.model.interfaces.FieldModel;
import kz.sgq.fs_tiktaktoe.mvp.model.FieldModelImpl;
import kz.sgq.fs_tiktaktoe.mvp.view.FieldView;

@InjectViewState
public class FieldPresenter extends MvpPresenter<FieldView> {
    private FieldModel model;

    public FieldPresenter(boolean mode, String stepText,
                          String playerOne, String playerTwo) {
        model = new FieldModelImpl(playerOne, playerTwo);
        init(mode, stepText);
    }

    private void init(boolean mode, String stepText) {
        model.setMode(mode);
        model.setStepText(stepText);
        getViewState().setStep(model.getStep());
        getViewState().setVictoryOne(model.getScorePlayerOne());
        getViewState().setVictoryTwo(model.getScorePlayerTwo());
    }


    public void onClickBox(int i) {
        if (model.isMode())
            onClickOne(i);
        else
            onClickTwo(i);
    }

    private void onClickOne(int i) {
        if (!model.getEnd() &&
                !model.checkDraw()) {
            if (model.onClickBox(i)) {
                getViewState().setListBox(i, "X");
                if (!model.checkVictory()) {
                    model.setStep(false);
                    if (model.checkDraw()) {
                        return;
                    }
                    int autoClick = model.stepBot();
                    model.onClickBox(autoClick);
                    getViewState().setListBox(autoClick, "O");
                    if (!model.checkVictory()) {
                        model.setStep(true);
                    } else {
                        getViewState().textColorEnd(model.getBoxWinList());
                        getViewState().setVictoryTwo(model.getScorePlayerTwo());
                    }
                } else
                    getViewState().textColorEnd(model.getBoxWinList());
                getViewState().setVictoryOne(model.getScorePlayerOne());
            }
        } else {
            getViewState().textColor();
            clearField();
        }
    }

    private void onClickTwo(int i) {
        if (!model.getEnd() &&
                !model.checkDraw()) {
            if (model.onClickBox(i)) {
                if (model.isStep()) {
                    getViewState().setListBox(i, "X");
                    if (model.checkVictory()) {
                        getViewState().textColorEnd(model.getBoxWinList());
                        getViewState().setVictoryOne(model.getScorePlayerOne());
                    } else {
                        model.setStep(false);
                    }
                } else {
                    getViewState().setListBox(i, "O");
                    if (model.checkVictory()) {
                        getViewState().textColorEnd(model.getBoxWinList());
                        getViewState().setVictoryTwo(model.getScorePlayerTwo());
                    } else {
                        model.setStep(true);
                    }
                }
                getViewState().setStep(model.getStep());
            }
        } else {
            getViewState().textColor();
            clearField();
        }
    }

    public void onClickBack() {
        getViewState().finishActivity();
    }


    public void onClickClear() {
        clearField();
        getViewState().setStep(model.getStep());
        model.clearScore();
        getViewState().setVictoryOne(model.getScorePlayerOne());
        getViewState().setVictoryTwo(model.getScorePlayerTwo());
        getViewState().textColor();
    }

    private void clearField() {
        getViewState().clearField();
        model.createLists();
        model.setStep(true);
    }

}

File: Kotlin/app/src/main/java/kz/sgq/kotlintictactoe/mvp/presenter/FieldPresenter.kt
package kz.sgq.kotlintictactoe.mvp.presenter

import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import kz.sgq.kotlintictactoe.mvp.model.FieldModelImpl
import kz.sgq.kotlintictactoe.mvp.model.interfaces.FieldModel
import kz.sgq.kotlintictactoe.mvp.view.FieldView

@InjectViewState
class FieldPresenter(
        mode: Boolean,
        stepText: String,
        playerOne: String,
        playerTwo: String
) : MvpPresenter<FieldView>() {

    private var model: FieldModel = FieldModelImpl(playerOne, playerTwo)

    init {
        model.setMode(mode)
        model.setStepText(stepText)
        viewState.setStep(model.getStep())
        viewState.setVictoryOne(model.getScorePlayerOne())
        viewState.setVictoryTwo(model.getScorePlayerTwo())
    }

    private fun clearField() {
        viewState.clearField()
        model.createLists()
        model.setStep(true)
    }

    fun onClickBox(i: Int) {
        if (model.isMode())
            onClickOne(i)
        else
            onClickTwo(i)
    }

    private fun onClickOne(i: Int) {
        if (!model.getEnd() && !model.checkDraw()) {
            if (model.onClickBox(i)) {
                viewState.setListBox(i, "X")
                if (!model.checkVictory()) {
                    model.setStep(false)
                    if (model.checkDraw()) {
                        return
                    }
                    val autoClick = model.stepBot()
                    model.onClickBox(autoClick)
                    viewState.setListBox(autoClick, "O")
                    if (!model.checkVictory()) {
                        model.setStep(true)
                    } else {
                        viewState.textColorEnd(model.getBoxWinList())
                        viewState.setVictoryTwo(model.getScorePlayerTwo())
                    }
                } else
                    viewState.textColorEnd(model.getBoxWinList())
                viewState.setVictoryOne(model.getScorePlayerOne())
            }
        } else {
            viewState.textColor()
            clearField()
        }
    }

    private fun onClickTwo(i: Int) {
        if (!model.getEnd() && !model.checkDraw()) {
            if (model.onClickBox(i)) {
                if (model.isStep()) {
                    viewState.setListBox(i, "X")
                    if (model.checkVictory()) {
                        viewState.textColorEnd(model.getBoxWinList())
                        viewState.setVictoryOne(model.getScorePlayerOne())
                    } else {
                        model.setStep(false)
                    }
                } else {
                    viewState.setListBox(i, "O")
                    if (model.checkVictory()) {
                        viewState.textColorEnd(model.getBoxWinList())
                        viewState.setVictoryTwo(model.getScorePlayerTwo())
                    } else {
                        model.setStep(true)
                    }
                }
                viewState.setStep(model.getStep())
            }
        } else {
            viewState.textColor()
            clearField()
        }
    }

    fun onClickBack() {
        viewState.finishActivity()
    }

    fun onClickClear() {
        clearField()
        viewState.setStep(model.getStep())
        model.clearScore()
        viewState.setVictoryOne(model.getScorePlayerOne())
        viewState.setVictoryTwo(model.getScorePlayerTwo())
        viewState.textColor()
    }
}